lib/spack/spack/spec.py

```diff
@@ -1684,17 +1905,17 @@ def to_node_dict(self, hash=ht.dag_hash):
             d['package_hash'] = package_hash
 
         # Note: Relies on sorting dict by keys later in algorithm.
-        deps = self.dependencies_dict(deptype=hash.deptype)
-
+        deps = self._dependencies_dict(deptype=hash.deptype)
         if deps:
             deps_list = []
-            for name, dspec in sorted(deps.items()):
+            for name, edges_for_name in sorted(deps.items()):
                 name_tuple = ('name', name)
-                hash_tuple = (hash.name, dspec.spec._cached_hash(hash))
-                type_tuple = ('type', sorted(str(s) for s in dspec.deptypes))
-                deps_list.append(syaml.syaml_dict([name_tuple,
-                                                   hash_tuple,
-                                                   type_tuple]))
+                for dspec in edges_for_name:
```

Comment:
I am not sure that it makes sense to write this fully generally, as we can *still* never have two direct dependencies with the same name.  I think right here, we should really be asserting that there is only one thing in the list, then taking the first element, instead of adding a loop nest.

I am having a hard time thinking of when we would generalize enough to have multiple direct dependencies *of the same name*, and I worry that we've generalized the interface for the uncommon case of traversing dependents, instead of keeping it simple for the much more common case of traversing dependencies.  I worry that will lead to weird errors.

Reply:
Resolving this per later discussion